Fix CLI port argument bug.

paweljakubas opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->
#401 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have added integration tests illustrating the problem
- [x] I have changed server by introducing PortOption datatype and introduced logic decided by team discussion
- [x] I have adopted exe cli


# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

paweljakubas committed:
introduce PortOption datatype and tweak logic in api server

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -139,15 +140,19 @@ import qualified Data.Text as T
 
+
+data PortOption = RandomPort | ExactPort (Maybe Warp.Port)
+    deriving Show
```
By using the `Maybe` type here, I think it's not obvious what the semantics are if we specify `ExactPort Nothing`.

Perhaps we can make this **clearer** by using the following options:
```hs
-- | Specifies a port on which the application server will listen.
data PortOption
    = DefaultPort
        -- ^ Listen on the default port (see 'defaultPort').
    | RandomPort
        -- ^ Select an unused port at random on which to listen.
    | SpecificPort Warp.Port
        -- ^ Listen on a specific port. 
    deriving Show
```
And then separately define:
```hs
-- | The default port on which to listen. The application server will listen
--   on this port if 'start' is called with 'DefaultPort'.
defaultPort :: Warp.Port
defaultPort = 9080
```
What do you think?

KtorZ replied:
This has been addressed :+1:

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -170,17 +175,21 @@ startOnSocket settings socket wl = Warp.runSettingsSocket settings socket
+        ExactPort portM -> case portM of
+            Just port -> (port,) <$> bindPortTCP port hostPreference
+            Nothing ->
+                (\port -> (port,) <$> bindPortTCP port hostPreference) 9080
```
I think this hard-coded value `9080` is a [magic number](https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#prefer-named-constants-over-magic-numbers).

In particular, it's hard-coded at least **twice**:
1. within the code;
2. within the preceding comment.

If one of these hard-coded values changes, we might forget to update the other one. And even worse, any code that **tests** this code will have to hard-code the **same** value, since there's no named constant to import and refer to.

What would you think about changing this to:
```hs
    acquire = case portOpt of
        DefaultPort ->
            (defaultPort,) <$> bindPortTCP defaultPort hostPreference
        ExactPort port ->
            (port,) <$> bindPortTCP port hostPreference
        RandomPort ->
            bindRandomPortTCP hostPreference
```
Then, separately, we can define:
```hs
-- | The default port on which to listen. The application server will listen
--   on this port if 'start' is called with 'DefaultPort'.
defaultPort :: Warp.Port
defaultPort = 9080
```
The comment to `withListeningSocket` could then be simplified to something like:
```hs
-- | Run an action with a TCP socket bound to a port, based on the specified
--   'PortOption'.
```
And for test code, we can simply import the constant.

What do you think?

rvl replied:
I have fixed some of the port number duplication.
The default port comes from the docopt text.
But I don't think magic numbers are too bad in tests.
Also it's often good not to share constants between the app and tests.

jonathanknowles replied:
I accept the argument that sometimes it's not good to share constants between the app and tests.

However, if a constant that appears in a test is hard-coded many times within the test suite, and if every instance of that constant is supposed to be the same, then I still think it would be better to use a named constant (that is not shared with the app). Otherwise, readers of the code are left wondering where the number came from.

jonathanknowles commented on lib/core/test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -762,6 +769,8 @@ createWalletViaCLI args mnemonics secondFactor passphrase = do
         return (c, T.unpack out, err)
 
+
+
```
Here there should only be one blank line, rather than three.

https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -139,15 +140,19 @@ import qualified Data.Text as T
 import qualified Data.Text.Encoding as T
 import qualified Network.Wai.Handler.Warp as Warp
 
+
```
https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions

jonathanknowles commented on lib/core/test/integration/Test/Integration/Scenario/CLI/Wallets.hs:
```diff
@@ -48,13 +49,13 @@ spec = do
-            T.unpack err `shouldContain` "Ok.\n"
-            out `shouldContain` walletName
+            checkSimpleCreateWallet
+                (Just "1337")
```
What is the relevance of the number `1337`? Is it an arbitrarily-chosen number, or should it be the same as a number somewhere else?

KtorZ replied:
Let's avoid the hard-coded number here indeed. Or at the very least, avoid repeating it through the whole file :pray:

jonathanknowles commented on lib/core/test/integration/Test/Integration/Scenario/CLI/Wallets.hs:
```diff
@@ -63,7 +64,7 @@ spec = do
             let passphrase = "Secure passphrase"
 
-            (c, out, err) <- createWalletViaCLI [walletName] mnemonics mnemonics2 passphrase
+            (c, out, err) <- createWalletViaCLI [walletName] mnemonics mnemonics2 passphrase (Just "1337")
```
This line is [too long](https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters).

How about:
```hs
            (c, out, err) <- createWalletViaCLI
                [walletName] mnemonics mnemonics2 passphrase (Just "1337")
```

jonathanknowles commented on lib/core/test/integration/Test/Integration/Scenario/CLI/Wallets.hs:
```diff
@@ -78,7 +79,7 @@ spec = do
             let passphrase = "Secure passphrase"
 
-            (c, out, err) <- createWalletViaCLI [walletName] mnemonics mnemonics2 passphrase
+            (c, out, err) <- createWalletViaCLI [walletName] mnemonics mnemonics2 passphrase (Just "1337")
```
Also here! :) https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

jonathanknowles commented on lib/core/test/integration/Test/Integration/Scenario/CLI/Wallets.hs:
```diff
@@ -90,7 +91,7 @@ spec = do
             let passphrase = "Secure passphrase"
 
-            (c, out, err) <- createWalletViaCLI [walletName] mnemonics mnemonics2 passphrase
+            (c, out, err) <- createWalletViaCLI [walletName] mnemonics mnemonics2 passphrase (Just "1337")
```
and here! :) https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

jonathanknowles commented on lib/core/test/integration/Test/Integration/Scenario/CLI/Wallets.hs:
```diff
@@ -99,6 +100,27 @@ spec = do
+        let walletName = "Wallet created via CLI"
+        checkSimpleCreateWallet
+            (Just "1338")
+            "15"
```
What is the significance of the number `15`? Is it the minimum mnemonic length? Perhaps we can introduce a named constant `minimumMnemonicLength`? (But then again, perhaps this is already defined elsewhere?)

jonathanknowles commented on lib/core/test/integration/Test/Integration/Scenario/CLI/Wallets.hs:
```diff
@@ -99,6 +100,27 @@ spec = do
+            (`shouldContain` "does not exist (Connection refused))\n")
+            (`shouldNotContain` walletName)
+
+
```
https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions

jonathanknowles commented on lib/core/test/integration/Test/Integration/Scenario/CLI/Wallets.hs:
```diff
@@ -142,7 +164,52 @@ spec = do
+    it "PORTS_01 - Can create wallet with specifying port" $ \_ -> do
+        let walletName = "Wallet created via CLI"
+        checkSimpleCreateWallet
+            (Just "9080")
```
Perhaps this can be:
```hs
        checkSimpleCreateWallet
            (Just $ show defaultPort)
```
If that is what you intended?

rvl replied:
Yes

jonathanknowles commented on lib/core/test/integration/Test/Integration/Scenario/CLI/Wallets.hs:
```diff
@@ -142,7 +164,52 @@ spec = do
+    -> (String -> IO ())
+    -> (String -> IO ())
+    -> IO ()
+checkSimpleCreateWallet portM mnemonicSize walletName exitCodeCheck errCheck outCheck = do
```
https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

jonathanknowles commented on lib/core/test/integration/Test/Integration/Scenario/CLI/Wallets.hs:
```diff
@@ -142,7 +164,52 @@ spec = do
+checkSimpleCreateWallet portM mnemonicSize walletName exitCodeCheck errCheck outCheck = do
+    Stdout mnemonics <- generateMnemonicsViaCLI ["--size", mnemonicSize]
+    let passphrase = "Secure passphrase"
+    (c, out, err) <- createWalletViaCLI [walletName] mnemonics "\n" passphrase portM
```
https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

jonathanknowles commented on lib/http-bridge/test/integration/Main.hs:
```diff
@@ -69,13 +71,15 @@ main = hspec $ do
     describe "Cardano.Wallet.HttpBridge.NetworkSpec" HttpBridge.spec
     describe "CLI commands not requiring bridge" MnemonicsCLI.spec
-    beforeAll startCluster $ afterAll killCluster $ after tearDown $ do
+    beforeAll (startCluster (Just 1337)) $ afterAll killCluster $ after tearDown $ do
```
Is this `1337` required to be the same number as the previous test cases? If so, perhaps can they all be the same named constant?

rvl replied:
Fixed, most tests use random port selection, and there is a defaultPort variable for the fixed port tests.

jonathanknowles commented on lib/http-bridge/test/integration/Main.hs:
```diff
@@ -124,9 +128,11 @@ main = hspec $ do
-        let baseURL = "http://localhost:1337/"
+        let baseURL = case portM of
+                Just port -> "http://localhost:" <> T.pack (show port) <> "/"
+                Nothing -> "http://localhost:9080/"
```
Perhaps this can be:
```hs
 "http://localhost:" <> show defaultPort <> "/"
```
I'm assuming that your intention is for this `9080` to be the **same value** as the one used in `Wallet.Api.Server`? (Feel free to correct my assumption.)

rvl replied:
Fixed, the server now returns the port it has used.

jonathanknowles reviewed (dismissed):
Hi @paweljakubas 

Thanks for looking at this issue!

I think this PR is generally good, though I also think there are some areas that we should fix. I've highlighted these areas in comments, along with some suggestions. I'd be interested to hear your thoughts.

Cheers!

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -170,17 +179,18 @@ startOnSocket settings socket wl = Warp.runSettingsSocket settings socket
--- then an unused port will be selected at random.
+-- | Run an action with a TCP socket bound to a port. If random port is chosen
+-- then an unused port will be selected at random. If exact port is chosen with
+-- no port specified then 9080 will be used.
```
This comment is now out of date.

rvl replied:
Thanks

KtorZ commented on exe/launcher/Main.hs:
```diff
@@ -91,13 +92,17 @@ main = do
+    let mWalletPort =
+            if args `isPresent` longOption "wallet-server-random-port"
+            then Nothing
+            else Just walletPort
```
It'll be best if the CLI fails if both `--wallet-server-port` and `--wallet-server-random-port` are passed. It should be one or the other. Which could probably be represented in the docopt specification directly by doing something like: 

```
cardano-wallet-launcher [options] (--wallet-server-random-port | --wallet-server-port=<PORT>)
```

KtorZ commented on exe/launcher/Main.hs:
```diff
@@ -118,16 +123,16 @@ nodeHttpBridgeOn stateDir port net =
-        ] ++ dbArg
+        ] ++ portArg ++ bridgePortArg ++ dbArg
+    portArg = maybe ["--random-port"] (\p -> ["--port", T.unpack (toText p)]) wp
+    bridgePortArg = ["--bridge-port", T.unpack (toText np)]
```
```suggestion
  args = mconcat
      [ [ "server" ]
      , [ "--network", if net == "local" then "testnet" else net ]
      , maybe ["--random-port"] (\p -> ["--port", showT p]) wp 
      , [ "--bridge-port", showT np ]
      ]
  showT = T.unpack . toText
```

KtorZ commented on exe/wallet/Main.hs:
```diff
@@ -140,7 +140,7 @@ and can be run "offline". (e.g. 'generate mnemonic')
 
 Usage:
-  cardano-wallet server [--network=STRING] [--port=INT] [--bridge-port=INT] [--database=FILE]
+  cardano-wallet server [--network=STRING] [--port=INT] [--random-port] [--bridge-port=INT] [--database=FILE]
```
Same here about `--port=INT` and `--random-port` being mutually exclusive.

KtorZ commented on lib/core/test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -742,12 +742,19 @@ createWalletViaCLI
+                Just port ->
+                    [ "exec", "--", "cardano-wallet"
+                    , "wallet", "create", "--port", port
+                    ] ++ args
```
```suggestion
let portArg =
        maybe [] (\p -> ["--port", p]) portM
let fullArgs = 
        [ "exec", "--", "cardano-wallet", "wallet", "create" ]
        ++ portArg ++ args
```

KtorZ commented on lib/http-bridge/test/integration/Main.hs:
```diff
@@ -172,13 +184,16 @@ main = hspec $ do
+            db <- MVar.newDBLayer
+            let tl = HttpBridge.newTransactionLayer
+            wallet <- newWalletLayer nullTracer block0 db nl tl
+            Server.start (putMVar port) walletListen wallet
```
That's a good idea and a good way to test that the random port binding works as expected :tada:

KtorZ commented on lib/http-bridge/test/integration/Main.hs:
```diff
@@ -158,6 +168,8 @@ main = hspec $ do
         -- some places where it's less annoying.
         (UseHandle h)
 
+    mkBaseUrl port = "http://localhost:" <> T.pack (show port) <> "/"
```
You probably need `toText` here instead of `show` since we have `port :: Port` I believe.

KtorZ commented on lib/http-bridge/test/integration/Main.hs:
```diff
@@ -71,13 +77,17 @@ main = hspec $ do
     describe "Cardano.Wallet.HttpBridge.NetworkSpec" HttpBridge.spec
     describe "CLI commands not requiring bridge" MnemonicsCLI.spec
-    beforeAll startCluster $ afterAll killCluster $ after tearDown $ do
+    beforeAll (startCluster ListenOnRandomPort) $ afterAll killCluster $ after tearDown $ do
```
This shouldn't be a problem for the API spec as long as we sort out the baseUrl in the manager, but the CLI specs also need to know about the port; So we probably gotta add it to the `Context` I believe.

paweljakubas replied:
I change it from `IO ()` to `IO Warp.Port` in start Server in `Cardano.Wallet.Api.Server` now we need to accommodate it in integration tests. And we will be able to use it when random port mode is ON in integration tests

KtorZ commented on lib/http-bridge/test/integration/Main.hs:
```diff
@@ -71,13 +77,17 @@ main = hspec $ do
+
+    let defaultPort = 8090
+    beforeAll (startCluster (ListenOnPort defaultPort)) $ afterAll killCluster $ after tearDown $ do
+        describe "Port tests using wallets CLI" (WalletsCLI.specPorts defaultPort)
```
What's the rationale for making this one completely separated ? It's quite heavy to start a whole new cluster. Perhaps what we really need here is simply to start another wallet server listening on given port. That can be tested without requiring a new cluster.

piotr-iohk replied:
+1. That'd also, sort of, test the case of two wallet's attached to a single server.

KtorZ commented on lib/core/test/integration/Test/Integration/Scenario/CLI/Wallets.hs:
```diff
@@ -142,7 +164,52 @@ spec = do
+    -> String
+    -> (ExitCode -> IO ())
+    -> (String -> IO ())
+    -> (String -> IO ())
```
not found of this helper, it's really "stringly-typed" and is more cryptic than readable IMO. I'd rather inline it in all the test above as it reads better inlined.

jonathanknowles commented on exe/launcher/Main.hs:
```diff
@@ -118,16 +123,16 @@ nodeHttpBridgeOn stateDir port net =
     networkArg = maybe [] (\d -> ["--networks-dir", d]) stateDir
 
-walletOn :: Maybe FilePath -> Port "Wallet" -> Port "Node" -> String -> Command
+walletOn :: Maybe FilePath -> Maybe (Port "Wallet") -> Port "Node" -> String -> Command
```
This line is now too long (https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters).

Suggest: 
```suggestion
walletOn
    :: Maybe FilePath
    -> Maybe (Port "Wallet")
    -> Port "Node"
    -> String
    -> Command
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -170,17 +179,17 @@ startOnSocket settings socket wl = Warp.runSettingsSocket settings socket
 
--- | Run an action with a TCP socket bound to a port. If no port is specified,
+-- | Run an action with a TCP socket bound to a port. If random port is chosen
 -- then an unused port will be selected at random.
```
This comment sounds a bit tautological. Suggest:
```suggestion
-- | Run an action with a TCP socket bound to a port specified by the `Listen`
--   parameter.
```

jonathanknowles commented on lib/http-bridge/test/integration/Main.hs:
```diff
@@ -71,13 +77,17 @@ main = hspec $ do
         describe "Addresses CLI tests" AddressesCLI.spec
+
+    let defaultPort = 8090
+    beforeAll (startCluster (ListenOnPort defaultPort)) $ afterAll killCluster $ after tearDown $ do
```
Let's try to keep lines to the length specified by https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

piotr-iohk commented on lib/core/test/integration/Test/Integration/Scenario/CLI/Wallets.hs:
```diff
@@ -142,7 +167,52 @@ spec = do
+    (c, out, err) <- createWalletViaCLI [walletName] mnemonics "\n" passphrase portM
+    exitCodeCheck c
+    errCheck (T.unpack err)
+    outCheck out
```
I understand that this check just landed here in this method, because it was there in the code in the first place :) Actually it is quite poor and I was planning to add more detailed checks (like here: https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/test/integration/Test/Integration/Scenario/CLI/Transactions.hs#L92-L97) but didn't get there yet for `Wallets.sh`. (ofc that's not in scope of this pr)

For that matter, I'd be rather for separating action and assertion not to be in a single method (but perhaps, that's just a personal preference)

KtorZ committed:
force launcher to use --random-port for wallet

Rationale here is to simplify a bit the requirement and busy work with the launcher that
we're going to remove / merge into the wallet cli soon.

KtorZ committed:
fix docopt port & random-port declaration

KtorZ committed:
Review tests for --port & fixes how the port get passed through tests

There's a bit of Haskell lens magic here. But that's for a greater good.

KtorZ committed:
add more tests about the --port to cover every command of the CLI

KtorZ committed:
re-generate nix packages

KtorZ reviewed (approved):
Changes LGTM. I've picked up the PR to resolve issues with the integration tests and reworked a bit the way integration scenario receive their corresponding port and made sure to thoroughly test every CLI commands.
